NDDataBase now (tries to) set the uncertainties parent_nddata attribute

MSeifert04 opened the pull request:
In #4152 I discovered that ``NDData`` togeter with the ``NDArithmeticMixin`` was not able to work with ``StdDevUncertainty`` as uncertainty. The problem was, that only the ``NDDataArray`` did set the ``parent_nddata`` attribute of the uncertainty while setting the uncertainty. Because the ``parent_nddata`` is necessary for ``StdDevUncertainty`` to work I added a few lines which at least try to set it in ``NDDataBase``.

Currently the ``except`` (when the uncertainty has no ``parent_nddata`` attribute) only makes a `pass` but it may also raise an Exception (which fails the current tests) or raise a warning (likewise) or be a case for the ``astropy.logging`` (but I only recently discovered this and I am not sure how to use it properly)

Since this is only an aside issue of the original issue I hope this does not close the original issue.

Any suggestions? And should I add a test case for the original problem (it's quite difficult, because all of ``nddata/mixins/tests/test_ndarithmetic.py``are based on ``NDDataArray`` not the different Mixins)

embray commented:
FYI this will probably end up in v1.0.6 rather than v1.0.5; we'll see.

MSeifert04 commented:
@embray: Thanks for adding the labels. Should I immediatly change my ``Change.rst`` text or wait for mwcraig? Also my written english seems a bit awkward to me, so if there is any proposed change I'll edit it.

mwcraig commented:
@MSeifert04 -- may as well wait. Just wanted to give you a heads up that it may be a day before I get to this, thanks for doing the fix!

mwcraig commented on CHANGES.rst:
```diff
@@ -655,6 +655,9 @@ Bug Fixes
     ``data`` as it should [#4137].
 
+  - ``NDDataBase`` now tries to set the ``parent_nddata`` attribute of the
+    uncertainty.
```
Could you please add the pull request number in brackets at the end of the changelog entry?

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -90,3 +90,9 @@ def uncertainty(self, value):
+        #Try setting the parent for the uncertainty. This is currently an
+        #enforced requirement for `StdDevUncertainty` for error propagation.
+        try:
+            self.uncertainty.parent_nddata = self
```
I think this should be `self._uncertainty.parent_nddate` (missing underscore before `uncertainty`)

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -90,3 +90,9 @@ def uncertainty(self, value):
+        #enforced requirement for `StdDevUncertainty` for error propagation.
+        try:
+            self.uncertainty.parent_nddata = self
+        except:
```
Please add a specific exception here (e.g. `except AttributeError:` if that is the error generated).

MSeifert04 replied:
That breaks a lot of existing test cases since we allow the uncertainty to be ``None`` which has no (parent_nddata attribute) but this could be fixed if I shift some lines around.

But thats unfortunatly not all: The ``test_nddata_base`` defines as minimal uncertainty something that has no ``parent_nddata`` attribute (https://github.com/astropy/astropy/blob/master/astropy/nddata/tests/test_nddata_base.py#L36). That made wonder what is considered the Minimal Uncertainty I should enforce here and I finally ended up with a ``try ... except pass`` which adds no restriction.

mwcraig commented:
@MSeifert04 -- it would be good to also add a test. If you want I can write one, but a good place to start would be with the code you wrote that raised the exception that led you to this fix. I would add the test to `test_nddata_base.py`.

I agree this is separate from #4152; in the long run we may need to rethink how to structure uncertainties, but for now this seems like a sensible way to go.

MSeifert04 commented:
@mwcraig : Thank you for looking over it. I changed the commented source. Later today I will check how to make the test cases. Wouldn't be bad if I at least try to do them, if they are bad I will come back to your offer.

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -89,4 +89,12 @@ def uncertainty(self, value):
+                raise AttributeError('Uncertainty must have an attribute '
+                                'parent_nddata.')
         self._uncertainty = value
+        
```
Apparently there is white space in this blank line...that is the only thing travis is complaining about right now :)

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -89,4 +89,12 @@ def uncertainty(self, value):
+            #enforced requirement for `StdDevUncertainty` for error propagation.
+            try:
+                value.parent_nddata = self
+            except:
```
Sorry wasn't clear about what I meant here. It is the except line itself that should have the name of the exception, like this:

```python
except AttributeError:
    pass
```

That way the *only* error this `except` allows to go without comment is an `AttributeError`; any other error generated by the line `value.parent_nddata = self` will (and should) be raised.

In this case it is a little hard to see what other error would be raised, but you could imagine someone subclassing `NDData`, making a setter for `uncertainty` and having a mistake in their code that generate some other error. We don't want to catch that error here, we just want an `AttributeError` raised if `value` has no attribute `parent_nddata`.

MSeifert04 commented:
Ok it took me a bit longer but I have finally added a few extra tests which seem to cover the changed lines. I was not sure if the added test with ``StdDevUncertainty`` was necessary but I wanted to cover the cases: 
1.) Has no parent_nddata attribute/setter and can't be set (uncertainty = None) 
2.) Has no parent_nddata attribute/setter but it can be set (uncertainty = MinimalUncertainty(5) ) 
3.) Has a specific setter for parent_nddata (uncertainty = StdDevUncertainty(None) ) - None was just out of curiosity.

Also I deleted the indentation for the empty line and changed the except-block. Thanks for all the comments.

MSeifert04 commented:
I was just wondering, reading through https://github.com/astropy/astropy-APEs/blob/master/APE7.rst why the uncertainty setter is implemented in ``NDDataBase`` and not in ``NDData`` (or what would be even better ``NDArithmeticMixin``) since NDDataBase should according to the APE enforce minimal restrictions (why force NDDataBase to use an uncertainty with an ``uncertainty_type`` attribute which is a string) even in NDData the documentation says we let the uncertainty be of any type (http://astropy.readthedocs.org/en/latest/api/astropy.nddata.NDData.html#astropy.nddata.NDData). Only for users who want arithmetics with uncertaintys (which would be the ``NDArithmeticMixin``) we should enforce the uncertainties to match our ``NDUncertainty``.

Or am I misunderstanding the APE (haven't read through all of the discussions).

embray commented:
> I was just wondering, reading through https://github.com/astropy/astropy-APEs/blob/master/APE7.rst why the uncertainty setter is implemented in NDDataBase and not in NDData (or what would be even better NDArithmeticMixin) since NDDataBase should according to the APE enforce minimal restrictions

I wasn't sure about this myself, but looking at the code it's simply to work around a particular shortcoming in how Python properties work.  In other words, it's just a technical workaround.  Basically, `NDDataBase` *defines* the common API for all NDData objects (just not the implementations).  However, in the interest of keeping the logical for the uncertainty property's *setter* in once place, it has to be implemented here.  *That said*, this can be worked around with a metaclass:  @mwcraig if you want I can work on a fix here so that `NDDataBase.uncertainty` is still an abstractproperty like the rest, but that still enforces the restrictions in `uncertainty.setter` on classes with *concrete* implementations of `uncertainty`.

> why force NDDataBase to use an uncertainty with an uncertainty_type attribute which is a string) even in NDData the documentation says we let the uncertainty be of any type (http://astropy.readthedocs.org/en/latest/api/astropy.nddata.NDData.html#astropy.nddata.NDData

I feel like it was probably a way to ensure that whatever the object *is*, it's something that's been designed to work in this framework.  In other words, saying "any type is allowed" was APE-7's way of punting on an exact description on how to handle uncertainties, which would be discussed later separately.  It was just declaring that there would be an interface for this.  I'm not sure what the entire rationale is behind the `uncertainty_type` attribute though.

> Only for users who want arithmetics with uncertaintys (which would be the NDArithmeticMixin) we should enforce the uncertainties to match our NDUncertainty

Not necessarily, if I understand correctly.  One could imagine use cases where we'd want to know for some data what the uncertainties are for each pixel, but any further arithmetic operations with that data are not well defined.  Or that maybe at least has different arithmetic that would be implemented through a different mixin than the standard `NDArithmeticMixin`.  But I'm just thinking out loud--I haven't been too involved in this package myself.

MSeifert04 commented:
Thanks for the answer but it leaves me more confused than before. If it is only a technical workaround why does the mask.setter work (where the abstract-property is defined in NDDataBase and the setter in NDData)? Is there any difference in how this works?

I was thinking about moving the `property uncertainty` from `NDDataBase` to `NDData` and instead define an `abstractproperty` in `NDDataBase`.
The specific `uncertainty.setter` would then be moved from `NDDataBase` to `NDArithmeticMixin`. This way the necessary checks for arithmetic continue to work but any subclass of `NDData` would not have this restrictions about having an attribute `uncertainty_type` which is a string. Instead such a subclass could simply use a `numpy.array` or an `astropy.quantity` or just a dict about gain/rdnoise. But if one wants to use the ArithmeticsMixin it has some more restrictions that are appropriate for uncertainty-propagation as implemented in `NDArithmeticMixin` (in fact another issue is that the NDArithmeticMixin also needs that the uncertainty implements the propagate-methods which are not enforced either in the init nor in the uncertainty setter so far).
If there are no stylistic/technical reasons against such a 'shuffle' I could implement the setter in `NDArithmeticMixin` so that there is no problem in actually using the arithmetics.

mwcraig commented:
>  However, in the interest of keeping the logical for the uncertainty property's setter in once place, it has to be implemented here. That said, this can be worked around with a metaclass: @mwcraig if you want I can work on a fix here so that NDDataBase.uncertainty is still an abstractproperty like the rest, but that still enforces the restrictions in uncertainty.setter on classes with concrete implementations of uncertainty.

@embray yes please :)

mwcraig commented:
@MSeifert04 -- @embray hit a lot of the highlights already, but IIRC, the desire in the discussions leading up to APE7 were to:

+ Ensure that if there was an uncertainty associated with an NDData-like object, it would be called `uncertainty`. @embray is right that the fact that the setter is implemented in `NDDataBase` is because I didn't know how to do it with a metaclass (I don't think there was a real design reason for it).
+ Enforce a requirement that there be some way ascertain what that uncertainty is; there were a couple people who felt strongly that a human-readable string was the best way to go. 

I think there is a use case for uncertainty that is unrelated to NDArithmeticMixin -- one where you want to provide uncertainty information even if you aren't providing a way to combine or propagate uncertainties.

I'll have some time on Friday to think through this in more detail but wanted to share some thoughts right away.

MSeifert04 commented:
I close this PR because it is completly integrated in my other PR #4234
